lib/srv/server/installer/autodiscover.go

```diff
@@ -273,16 +299,58 @@ func (ani *AutoDiscoverNodeInstaller) configureTeleportNode(ctx context.Context,
 			fmt.Sprintf(`--azure-client-id=%s`, shsprintf.EscapeDefaultContext(ani.AzureClientID)))
 	}
 
-	ani.Logger.InfoContext(ctx, "Writing teleport configuration", "teleport", ani.binariesLocation.Teleport, "args", teleportNodeConfigureArgs)
+	ani.Logger.InfoContext(ctx, "Generating teleport configuration", "teleport", ani.binariesLocation.Teleport, "args", teleportNodeConfigureArgs)
 	teleportNodeConfigureCmd := exec.CommandContext(ctx, ani.binariesLocation.Teleport, teleportNodeConfigureArgs...)
 	teleportNodeConfigureCmdOutput, err := teleportNodeConfigureCmd.CombinedOutput()
 	if err != nil {
 		return trace.Wrap(err, string(teleportNodeConfigureCmdOutput))
 	}
 
+	defer func() {
+		// If an error occurs before the os.Rename, let's remove the `.new` file to prevent any leftovers.
+		// Error is ignored because the file might be already removed.
+		_ = os.Remove(teleportYamlConfigurationPathNew)
+	}()
+
+	// Check if file already exists and has the same content that we are about to write
+	if _, err := os.Stat(teleportYamlConfigurationPath); err == nil {
+		hashExistingFile, err := checksum(teleportYamlConfigurationPath)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		hashNewFile, err := checksum(teleportYamlConfigurationPathNew)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		if hashExistingFile == hashNewFile {
+			return trace.AlreadyExists("teleport.yaml is up to date")
+		}
+	}
+
+	if err := os.Rename(teleportYamlConfigurationPathNew, teleportYamlConfigurationPath); err != nil {
+		return trace.Wrap(err)
+	}
+
 	return nil
 }
 
+func checksum(filename string) (string, error) {
+	f, err := utils.OpenFileAllowingUnsafeLinks(filename)
```

Comment:
I see that tctl already has a similar function, but the usage of `utils.OpenFileAllowingUnsafeLinks` does raise my eyebrow. Would it be possible to use `utils.OpenFileNoUnsafeLinks` instead? If not, then I think it'd be a good idea to leave a comment explaining why following symlinks is warranted in this case. I don't have enough context on this code to assess this.

Reply:
I'll change it to use `utils.OpenFileNoUnsafeLinks`.
This happens for two files and those files are written by `teleport node configure` which creates the files and no symlinks.
Also, we expect no manual intervention in these files because teleport is meant to be installed automatically.

Reply:
@ravicious @marcoandredinis is it really necessary to block symlinks here? this breaks the tests on my macbook where `/var` is a symlink. How do we know `/etc` won't be a symlink and why should we care?

IIRC the reason to avoid reading files from symlinks is to avoid reading them as root and leaking info about files owned by other users to an unprivileged user who created a malicious symlink

```
    --- FAIL: TestAutoDiscoverNode/gcp_adds_a_label_with_the_project_id (0.01s)
        autodiscover_test.go:369:
                Error Trace:    /Users/nic/src/teleport/lib/srv/server/installer/autodiscover_test.go:369
                Error:          Received unexpected error:
                                opening file /var/folders/x_/33kp49n52rg8hs97k5m96rnw0000gn/T/TestAutoDiscoverNodegcp_adds_a_label_with_the_project_id462103172/001/etc/teleport.yaml, symlink not allowed in path: /var
                Test:           TestAutoDiscoverNode/gcp_adds_a_label_with_the_project_id
```

Reply:
I'll change it back to `utils.OpenFileAllowingUnsafeLinks`
https://github.com/gravitational/teleport/pull/46600